File: src/query/tablet.js
import path from "path";
import { ReadableStream } from "@swimburger/isomorphic-streams";
import { isEmpty } from "../stream.js";
import { mow, sow } from "../record.js";
import { keyGroups } from "./groups.js";

export function makeStateInitial({ query, entry, thingQuerying }, tablet) {
  // in a querying tablet, set initial entry to the base of the tablet
  // and preserve the received entry for sowing grains later
  // if tablet base is different from previous entry base
  // sow previous entry into the initial entry
  const isSameBase = tablet.base === query._;

  const doDiscard = entry === undefined || isSameBase;

  const entryFallback = doDiscard ? { _: tablet.base } : entry;

  const doSow = !doDiscard;

  const entryInitial = doSow
    ? sow(
        { _: tablet.base },
        { _: entry._, [entry._]: entry[entry._] },
        tablet.base,
        entry._,
      )
    : entryFallback;

  const entryBaseChanged = entry === undefined || entry._ !== entryInitial._;

  // if entry base changed forget thingQuerying
  const thingQueryingInitial = entryBaseChanged ? undefined : thingQuerying;

  return { entry: entryInitial, query, thingQuerying: thingQueryingInitial };
}

/**
 * Match a key group against query grains.
 * Returns { matched, entry, query, thingQuerying } where matched is true
 * if any value in the group satisfied all grain predicates.
 */
function matchGroup(key, values, tablet, grains, stateInitial) {
  let groupEntry = { ...stateInitial.entry };
  let groupQuery = { ...stateInitial.query };
  let matched = false;
  let groupThingQuerying = undefined;

  for (const value of values) {
    const trait_ = tablet.traitIsFirst ? key : value;
    const thing = tablet.thingIsFirst ? key : value;

    const grainNew = {
      _: tablet.base,
      [tablet.trait]: trait_,
      [tablet.thing]: thing,
    };

    // all grains must match
    const isMatchGrains = grains.reduce((acc, grain) => {
      const isMatchGrain = tablet.traitIsRegex
        ? new RegExp(grain[tablet.trait]).test(trait_)
        : grain[tablet.trait] === trait_;

      return acc === undefined ? isMatchGrain : acc && isMatchGrain;
    }, undefined);

    // when parent key (thingQuerying) is set, also require it matches
    const doDiff = stateInitial.thingQuerying !== undefined;
    const isMatchQuerying = doDiff
      ? stateInitial.thingQuerying === thing
      : true;

    const isMatch = isMatchGrains && isMatchQuerying;

    if (isMatch) {
      matched = true;
      groupThingQuerying = thing;
      groupEntry = sow(groupEntry, grainNew, tablet.trait, tablet.thing);
      groupQuery = sow(groupQuery, grainNew, tablet.trait, tablet.thing);
    }
  }

  return { matched, entry: groupEntry, query: groupQuery, thingQuerying: groupThingQuerying };
}

export async function queryTabletStream(
  fs,
  dir,
  tablet,
  { query, entry, thingQuerying },
  first,
) {
  const filepath = path.join(dir, tablet.filename);
  const empty = await isEmpty(fs, filepath);

  // first tablet needs lines — empty file means no matches
  // later tablet avoids lines — empty file means match all
  if (empty && first) {
    return ReadableStream.from([]);
  }

  if (empty && !first) {
    return ReadableStream.from([{ query, entry, thingQuerying }]);
  }

  const stateInitial = makeStateInitial(
    { query, entry, thingQuerying },
    tablet,
  );

  const grains = mow(stateInitial.query, tablet.trait, tablet.thing);

  async function* matchedEntries() {
    for await (const { key, values } of keyGroups(fs, dir, tablet.filename)) {
      const result = matchGroup(key, values, tablet, grains, stateInitial);

      if (result.matched) {
        yield {
          entry: result.entry,
          query: result.query,
          thingQuerying: result.thingQuerying,
        };
      }
    }
  }

  return ReadableStream.from(matchedEntries());
}

File: src/dataset/query/tablet.rs
use super::groups::{key_groups, KeyGroup};
use super::line::State;
use super::strategy::Tablet;
use crate::{Dataset, Entry, Grain, Result};
use async_stream::try_stream;
use futures_core::stream::Stream;
use regex::Regex;

fn make_state_initial(state: &State, tablet: &Tablet) -> State {
    // if tablet base differs from query base, sow previous entry into a new entry
    let is_same_base = tablet.base == state.query.base;

    let do_discard = state.entry.is_none() || is_same_base;

    let entry_initial = if do_discard {
        Entry::new(&tablet.base)
    } else {
        let prev = state.entry.as_ref().expect("checked is_some via do_discard");
        let e = Entry::new(&tablet.base);

        let g = Grain {
            base: prev.base.clone(),
            base_value: prev.base_value.clone(),
            leaf: prev.base.clone(),
            leaf_value: None,
        };

        e.sow(&g, &tablet.base, &prev.base)
    };

    let entry_base_changed = match &state.entry {
        None => true,
        Some(e) => e.base != entry_initial.base,
    };

    // if entry base changed, forget parent join key
    let thing_querying_initial = if entry_base_changed {
        None
    } else {
        state.thing_querying.clone()
    };

    State {
        entry: Some(entry_initial),
        query: state.query.clone(),
        thing_querying: thing_querying_initial,
    }
}

/// Match a key group against query grains.
/// Returns Some(State) if any value in the group matched, None otherwise.
fn match_group(
    group: &KeyGroup,
    tablet: &Tablet,
    grains: &[Grain],
    state_initial: &State,
) -> Option<State> {
    let mut group_entry = state_initial.entry.clone()?;
    let mut group_query = state_initial.query.clone();
    let mut matched = false;
    let mut group_thing_querying: Option<String> = None;

    for value in &group.values {
        let trait_ = if tablet.trait_is_first {
            group.key.clone()
        } else {
            value.clone()
        };

        let thing = if tablet.thing_is_first {
            group.key.clone()
        } else {
            value.clone()
        };

        let grain_new = if tablet.thing_is_first {
            if tablet.trait_is_first {
                Grain {
                    base: tablet.thing.clone(),
                    base_value: Some(group.key.clone()),
                    leaf: tablet.thing.clone(),
                    leaf_value: None,
                }
            } else {
                Grain {
                    base: tablet.thing.clone(),
                    base_value: Some(group.key.clone()),
                    leaf: tablet.trait_.clone(),
                    leaf_value: Some(value.clone()),
                }
            }
        } else if tablet.trait_is_first {
            Grain {
                base: tablet.trait_.clone(),
                base_value: Some(group.key.clone()),
                leaf: tablet.thing.clone(),
                leaf_value: Some(value.clone()),
            }
        } else {
            Grain {
                base: tablet.base.clone(),
                base_value: None,
                leaf: tablet.base.clone(),
                leaf_value: None,
            }
        };

        // all grains must match
        let is_match_grains = grains.iter().fold(None, |acc, grain| {
            let re_str = if tablet.trait_is_first {
                grain.base_value.as_deref().unwrap_or("")
            } else {
                grain.leaf_value.as_deref().unwrap_or("")
            };

            let is_match_grain = if tablet.trait_is_regex {
                Regex::new(re_str)
                    .ok()
                    .map_or(false, |re| re.is_match(&trait_))
            } else {
                re_str == trait_
            };

            Some(match acc {
                None => is_match_grain,
                Some(prev) => prev && is_match_grain,
            })
        });

        // when parent join key is set, also require it matches
        let is_match_querying = match &state_initial.thing_querying {
            Some(tq) => *tq == thing,
            None => true,
        };

        let is_match = is_match_grains.unwrap_or(false) && is_match_querying;

        if is_match {
            matched = true;
            group_thing_querying = Some(thing);
            group_entry = group_entry.sow(&grain_new, &tablet.trait_, &tablet.thing);
            group_query = group_query.sow(&grain_new, &tablet.trait_, &tablet.thing);
        }
    }

    if matched {
        Some(State {
            entry: Some(group_entry),
            query: group_query,
            thing_querying: group_thing_querying,
        })
    } else {
        None
    }
}

pub fn query_tablet_stream(
    dataset: Dataset,
    tablet: Tablet,
    state: State,
    is_first_tablet: bool,
) -> impl Stream<Item = Result<State>> {
    try_stream! {
        let filepath = dataset.dir.join(&tablet.filename);
        let file_exists = std::fs::metadata(&filepath).is_ok();

        // first tablet needs lines — empty/missing file means no matches
        // later tablet — empty/missing file means match all
        if !file_exists {
            if !is_first_tablet {
                yield state;
            }
            return;
        }

        let state_initial = make_state_initial(&state, &tablet);
        let grains = state_initial.query.mow(&tablet.trait_, &tablet.thing);

        let groups = key_groups(&filepath)?;

        for group in &groups {
            if let Some(matched_state) = match_group(group, &tablet, &grains, &state_initial) {
                yield matched_state;
            }
        }
    }
}
